azurerm/helpers/authentication/auth_method_client_cert.go

```diff
@@ -0,0 +1,115 @@
+package authentication
+
+import (
+	"crypto/rsa"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/hashicorp/go-multierror"
+	"golang.org/x/crypto/pkcs12"
+)
+
+type servicePrincipalClientCertificateAuth struct {
+	clientId           string
+	clientCertPath     string
+	clientCertPassword string
+	environment        string
+	subscriptionId     string
+	tenantId           string
+}
+
+func newServicePrincipalClientCertificateAuth(b Builder) authMethod {
+	return servicePrincipalClientCertificateAuth{
+		clientId:           b.ClientID,
+		clientCertPath:     b.ClientCertPath,
+		clientCertPassword: b.ClientCertPassword,
+		environment:        b.Environment,
+		subscriptionId:     b.SubscriptionID,
+		tenantId:           b.TenantID,
+	}
+}
+
+func (a servicePrincipalClientCertificateAuth) getAuthorizationToken(oauthConfig *adal.OAuthConfig, endpoint string) (*autorest.BearerAuthorizer, error) {
+	certificateData, err := ioutil.ReadFile(a.clientCertPath)
+	if err != nil {
+		return nil, fmt.Errorf("Error reading Client Certificate %q: %v", a.clientCertPath, err)
+	}
+
+	// Get the certificate and private key from pfx file
+	certificate, rsaPrivateKey, err := decodePkcs12(certificateData, a.clientCertPassword)
+	if err != nil {
+		return nil, fmt.Errorf("Error decoding pkcs12 certificate: %v", err)
+	}
+
+	spt, err := adal.NewServicePrincipalTokenFromCertificate(*oauthConfig, a.clientId, certificate, rsaPrivateKey, endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	err = spt.Refresh()
+	if err != nil {
+		return nil, err
+	}
+
+	auth := autorest.NewBearerAuthorizer(spt)
+	return auth, nil
+}
+
+func (a servicePrincipalClientCertificateAuth) validate() error {
+	var err *multierror.Error
+
+	fmtErrorMessage := "A %s must be configured when authenticating as a Service Principal using a Client Certificate."
+
+	if a.subscriptionId == "" {
+		err = multierror.Append(err, fmt.Errorf(fmtErrorMessage, "Subscription ID"))
+	}
+
+	if a.clientId == "" {
+		err = multierror.Append(err, fmt.Errorf(fmtErrorMessage, "Client ID"))
+	}
+
+	if a.clientCertPath == "" {
+		err = multierror.Append(err, fmt.Errorf(fmtErrorMessage, "Client Certificate Path"))
+	} else {
+		if strings.HasSuffix(strings.ToLower(a.clientCertPath), ".pfx") {
```

Comment:
Is there a reason to enforce that the file ends with .pfx? I've in the past used things like `cert.pfx.old`

Reply:
this is intentional since we require a PFX file here (whereas other providers require `cer` files etc) - so I'd suggest we start out strict and we can loosen this if required